libtransmission/peer-mgr-wishlist.cc

```diff
@@ -134,12 +134,20 @@ std::vector<tr_block_span_t> Wishlist::next(Wishlist::Mediator const& mediator,
         return {};
     }
 
-    // We usually won't need all the candidates until endgame, so don't
-    // waste cycles sorting all of them here. partial sort is enough.
     auto candidates = getCandidates(mediator);
-    auto constexpr MaxSortedPieces = size_t{ 30 };
-    auto const middle = std::min(std::size(candidates), MaxSortedPieces);
-    std::partial_sort(std::begin(candidates), std::begin(candidates) + middle, std::end(candidates));
+
+    if (mediator.isSequentialDownload())
+    {
+        // no sorting - sequential order
```

Comment:
This is a good start but I wonder a little bit about this -- bypassing the sort means that we're no longer to going to prioritize nearly-complete pieces (useful for spotting any peers that send corrupt data) or high-priority pieces/files (useful for users who have clicked "priority: high" on a file).

For that reason I feel like we ought to remove this `if ... else` change and keep the original code that calls `std::partial_sort()`.

Instead, let's make `Candidate::salt` a `tr_piece_index_t` instead of a `uint8_t` and populate it sequentially instead of randomly iff `isSequentialDownload()` is true. That way,  `Candidate.compare()` will return a sequential piece sort based on the code [here](https://github.com/transmission/transmission/blob/44546674d955c9b2982ec3c2b18d547142b0c9e4/libtransmission/peer-mgr-wishlist.cc#L52).

```diff
diff --git a/libtransmission/peer-mgr-wishlist.cc b/libtransmission/peer-mgr-wishlist.cc
index 6c83e43e4..e194f6b3c 100644
--- a/libtransmission/peer-mgr-wishlist.cc
+++ b/libtransmission/peer-mgr-wishlist.cc
@@ -19,15 +19,16 @@
 
 namespace
 {
+using SaltType = tr_piece_index_t;
 
 struct Candidate
 {
     tr_piece_index_t piece;
     size_t n_blocks_missing;
     tr_priority_t priority;
-    uint8_t salt;
+    SaltType salt;
 
-    Candidate(tr_piece_index_t piece_in, size_t missing_in, tr_priority_t priority_in, uint8_t salt_in)
+    Candidate(tr_piece_index_t piece_in, size_t missing_in, tr_priority_t priority_in, SaltType salt_in)
         : piece{ piece_in }
         , n_blocks_missing{ missing_in }
         , priority{ priority_in }
@@ -86,14 +87,16 @@ std::vector<Candidate> getCandidates(Wishlist::Mediator const& mediator)
     }
 
     // transform them into candidates
-    auto salter = tr_salt_shaker{};
+    auto salter = tr_salt_shaker<SaltType>{};
     auto const n = std::size(wanted_pieces);
     auto candidates = std::vector<Candidate>{};
+    auto const is_sequential = mediator.isSequentialDownload();
     candidates.reserve(n);
     for (size_t i = 0; i < n; ++i)
     {
         auto const [piece, n_missing] = wanted_pieces[i];
-        candidates.emplace_back(piece, n_missing, mediator.priority(piece), salter());
+        auto const salt = is_sequential ? piece : salter();
+        candidates.emplace_back(piece, n_missing, mediator.priority(piece), salt);
     }
 
     return candidates;
```

Reply:
sounds good to me ! and thanks for the patch :)